vowpalwabbit/csoaa.cc

```diff
@@ -129,7 +145,69 @@ void predict_or_learn(csoaa& c, single_learner& base, example& ec)
   ec.pred.multiclass = prediction;
 }
 
-void finish_example(vw& all, csoaa&, example& ec) { COST_SENSITIVE::finish_example(all, ec); }
+template <bool probabilities>
+void finish_example(vw& all, csoaa& c, example& ec)
+{
+  if (probabilities)
+  {
+    // === Compute multiclass_log_loss
+    // TODO:
+    // What to do if the correct label is unknown, i.e. (uint32_t)-1?
+    //   Suggestion: increase all.sd->weighted_unlabeled_examples???,
+    //               but not sd.example_number, so the average loss is not influenced.
+    // What to do if the correct_class_prob==0?
+    //   Suggestion: have some maximal multiclass_log_loss limit, e.g. 999.
+    float multiclass_log_loss = 999;  // -log(0) = plus infinity
+    float correct_class_prob = 0;
+    if (ec.l.multi.label <= c.num_classes)  // prevent segmentation fault if labeĺ==(uint32_t)-1
```

Comment:
Is uint32_t a special value?

Reply:
There does seem to be a mismatch, I can take this comment out